src/dataflow/src/render/context.rs

```diff
@@ -115,301 +99,384 @@ where
-            // If we found such a key, use a more advanced implementation.
-            if let Some((key, row)) = constrained_keys.pop() {
-                let (_id, oks, errs) = &trace[&key];
-                let result = Self::flat_map_ref_inner(oks, row, logic);
-                Some((result, errs.as_collection(|k, _v| k.clone())))
-            } else {
-                let (_id, oks, errs) = trace.values().next().expect("Empty arrangement");
-                Some((
-                    oks.flat_map_ref(move |_k, v| logic(RefOrMut::Ref(&v))),
-                    errs.as_collection(|k, _v| k.clone()),
-                ))
+        }
+
+        // No key, or no matching arrangement (that would be weird).
+        // We should now prefer to use an arrangement if available (avoids allocation),
+        // and resort to using a collection if not available (copies all rows).
+        if let Some(flavor) = self.arranged.values().next() {
+            match flavor {
+                ArrangementFlavor::Local(oks, errs) => {
+                    let oks = Self::flat_map_core(&oks, None, logic);
+                    let errs = errs.as_collection(|k, _v| k.clone());
+                    (oks, errs)
+                }
+                ArrangementFlavor::Trace(_, oks, errs) => {
+                    let oks = Self::flat_map_core(&oks, None, logic);
+                    let errs = errs.as_collection(|k, _v| k.clone());
+                    (oks, errs)
+                }
             }
         } else {
-            None
+            use timely::dataflow::operators::Map;
+            let (oks, errs) = self.as_collection();
+            (
+                oks.inner
+                    .flat_map(move |(mut v, t, d)| logic(RefOrMut::Mut(&mut v), &t, &d)),
+                errs,
+            )
         }
     }
 
     /// Factored out common logic for using literal keys in general traces.
     ///
     /// This logic is sufficiently interesting that we want to write it only
     /// once, and thereby avoid any skew in the two uses of the logic.
-    fn flat_map_ref_inner<Tr, I, L>(
+    fn flat_map_core<Tr, I, L>(
         trace: &Arranged<S, Tr>,
-        row: V,
+        key: Option<V>,
         mut logic: L,
-    ) -> Collection<S, I::Item, Diff>
+    ) -> timely::dataflow::Stream<S, I::Item>
     where
         Tr: TraceReader<Key = V, Val = V, Time = S::Timestamp, R = repr::Diff> + Clone + 'static,
         Tr::Batch: BatchReader<V, Tr::Val, S::Timestamp, repr::Diff> + 'static,
         Tr::Cursor: Cursor<V, Tr::Val, S::Timestamp, repr::Diff> + 'static,
         I: IntoIterator,
         I::Item: Data,
-        L: FnMut(RefOrMut<V>) -> I + 'static,
+        L: for<'a, 'b> FnMut(RefOrMut<'b, V>, &'a S::Timestamp, &'a repr::Diff) -> I + 'static,
     {
-        use differential_dataflow::AsCollection;
+        let mode = if key.is_some() { "index" } else { "scan" };
+        let name = format!("ArrangementFlatMap({})", mode);
```

Comment:
@philip-stoev, this causes item no.2. The plan is the same, is just that the operator has been given a more concrete name.

Reply:
Thanks!